### PR TITLE
Index analysis using Hilbert space-filling curve

### PR DESCRIPTION
**Pull request by Artemmm91**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add support for index analysis over `hilbertEncode`.

If a table has hilbertEncode (aka Hilbert curve) in its index, e.g. `ORDER BY hilbertEncode(x, y)` then queries that use both `x` and `y` can be processed efficient, for example `WHERE 1 <= x AND x <= 2 AND 3 <= y AND y <= 4`. Works with the setting `analyze_index_with_space_filling_curves`.

Has about x2 time improvements comparing to Morton Curve on queries with small ranges, and around 15% improvement on queries with big ranges. 

Has around x2 time improvement on queries using both columns compared to simple lexicographic index (aka `ORDER BY (x, y)`)

Continuation of #60156. Continuation of #55642. 

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>CI Settings</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step
- [ ]  Allow: Integration Tests
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Unit tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All with aarch64
- [ ]  Allow: All with ASAN
- [ ]  Allow: All with TSAN
- [ ]  Allow: All with Analyzer
- [ ]  Allow: All with Azure
- [ ]  Allow: Add your option here
---
- [ ]  Exclude: Fast test
- [ ]  Exclude: Integration Tests
- [ ]  Exclude: Stateless tests
- [ ]  Exclude: Stateful tests
- [ ]  Exclude: Performance tests
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN
- [ ]  Exclude: All with MSAN
- [ ]  Exclude: All with UBSAN
- [ ]  Exclude: All with Coverage
- [ ]  Exclude: All with Aarch64
---
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)
- [ ]  allow: batch 1 for multi-batch jobs
- [ ]  allow: batch 2
- [ ]  allow: batch 3
- [ ]  allow: batch 4, 5 and 6
</details>

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit a7aa24a968fec07b72872881470878949c329738 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64662/a7aa24a968fec07b72872881470878949c329738/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

**Comment by Artemmm91** [comment]
Already did some benches (similar to ones in https://github.com/ClickHouse/ClickHouse/pull/55642)
```
CREATE TABLE hits_clientip_remoteip ENGINE = MergeTree ORDER BY (ClientIP, RemoteIP) AS (SELECTCAST(ClientIP, 'UInt32') AS ClientIP, CAST(RemoteIP, 'UInt32') AS RemoteIP, Title FROM datasets.hits_v1)
CREATE TABLE hits_morton ENGINE = MergeTree ORDER BY mortonEncode(ClientIP, RemoteIP) AS SELECTCAST(ClientIP, 'UInt32') AS ClientIP, CAST(RemoteIP, 'UInt32') AS RemoteIP, Title FROM datasets.hits_v1
CREATE TABLE hits_hilbert ENGINE = MergeTree ORDER BY hilbertEncode(ClientIP, RemoteIP) AS SELECT CAST(ClientIP, 'UInt32') AS ClientIP, CAST(RemoteIP, 'UInt32') AS RemoteIP, Title FROM datasets.hits_v1
```

1) Range select from lexigraphically ordered table
```
SELECT Title FROM hits_clientip_remoteip WHERE (ClientIP >= 2801131153) AND (ClientIP <= 2901131153) AND (Title != '') ORDER BY Title ASC LIMIT 10
Elapsed: 0.202 sec. Processed 180.22 thousand rows
```

```
SELECT Title FROM hits_clientip_remoteip WHERE (RemoteIP >= 2801131153) AND (RemoteIP <= 2901131153) AND (Title != '') ORDER BY Title ASC LIMIT 10
Elapsed: 0.992 sec. Processed 8.87 million rows
```

2) Range select from table ordered by Morton Curve
```
SELECT Title FROM hits_morton WHERE (ClientIP >= 2801131153) AND (ClientIP <= 2901131153) AND (Title != '') ORDER BY Title ASC LIMIT 10
Elapsed: 0.273 sec. Processed 630.78 thousand rows
```

```
SELECT Title FROM hits_morton WHERE (RemoteIP >= 2801131153) AND (RemoteIP <= 2901131153) AND (Title != '') ORDER BY Title ASC LIMIT 10
Elapsed: 0.275 sec. Processed 466.94 thousand rows
```

3) Range select from table ordered by Hilbert Curve
```
SELECT Title FROM hits_hilbert WHERE (ClientIP >= 2801131153) AND (ClientIP <= 2901131153) AND (Title != '') ORDER BY Title ASC LIMIT 10
Elapsed: 0.235 sec. Processed 393.22 thousand rows
```

```
SELECT Title FROM hits_hilbert WHERE (RemoteIP >= 2801131153) AND (RemoteIP <= 2901131153) AND (Title != '') ORDER BY Title ASC LIMIT 10

Elapsed: 0.281 sec. Processed 434.18 thousand rows
```

So results are as expected, and we can see some visible improvements in one of selects. I'll make more in-depth analysis is it just coincidence or on some cases Hilbert curve is really better than Morton

**Comment by Artemmm91** [comment]
Also i have some ideas on improvement - right now function hilbertIntervalToHyperrectangles (or mortonIntervalToHyperrectangles) is called on every granule. But it seems more efficient to call other function (that will transform range from querry to curve hyperrectangles) just once and for every granules just intersect rectangles

**Review by yariks5s — COMMENTED** [review]
2024.06.03 16:18:23.668195 [ 179136 ] {} <Fatal> BaseDaemon: (version 24.6.1.1, build id: 7833BE2358CC6B1C8D4F95EC709E8BD9D518F077, git hash: 6b2a9de2f1b37fe9e916ed78d534bd3fe8d0508a) (from thread 179086) (query_id: 1e112d6e-7d06-418f-88c0-76a054ac1055) (query: CREATE TABLE numbers1 ENGINE = Memory AS SELECT number FROM numbers(1000);) Received signal Segmentation fault (11)
2024.06.03 16:18:23.668240 [ 179136 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Address not mapped to object.

File: src/Storages/MergeTree/KeyCondition.cpp [hunk 1]
```diff
@@ -730,15 +735,15 @@ static NameSet getAllSubexpressionNames(const ExpressionActions & key_expr)
 
 void KeyCondition::getAllSpaceFillingCurves()
 {
-    /// So far the only supported function is mortonEncode (Morton curve).
```
Comment (yariks5s) [inline src/Storages/MergeTree/KeyCondition.cpp:733]: image

There is a lot of space-filling functions, so I think it's better to keep this comment and add Hilbert curve there

**Comment by yariks5s** [comment]
@Artemmm91 merging with master won't resolve this problem because these changes raised a Segmentation fault, above there is a comment with the possible repro, I can also check what's wrong if you want.

**Comment by Artemmm91** [comment]
> @Artemmm91 merging with master won't resolve this problem because these changes raised a Segmentation fault, above there is a comment with the possible repro, I can also check what's wrong if you want.

yeah, i see. it seems like the problem is in my code, i will search bugs in it then

**Comment by Artemmm91** [comment]
bruh, i already had this bug, but fixed it locally and forgot to commit

**Comment by Artemmm91** [comment]
Had a more in-depth analysis and made some benchmarks on hits datasets.

Generated a bunch of queries for ranges for RemoteIP field, ClientIP field and both fields in the same queries, and made 2 different benchmarks - the first is where ranges could be up to a whole interval of IPs (up to 1_000_000_000) and the second  one is where ranges are smaller (up to 1_000_000). 

And there are some plots of benchmarks. As you can see on big ranges there is improvement by near 15-20%, and in small ranges is up to x2 improvement compared to morton curve

big_ranges_big_granules_both
big_ranges_big_granules_client
big_ranges_big_granules_remote
small_ranges_big_granules_both
small_ranges_big_granules_client
small_ranges_big_granules_remote

File: src/Common/HilbertUtils.h [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        segmentBinaryPartition(finish_range.begin, finish, next_bits, callback);
+    }
+}
+
+// Given 2 points representing ends of the range of Hilbert Curve that lies in a whole domain.
+// The are neighbour corners of some square - and the function returns ranges of both sides of this square
+inline std::array<std::pair<UInt64, UInt64>, 2> createRangeFromCorners(UInt64 x1, UInt64 y1, UInt64 x2, UInt64 y2)
+{
+    UInt64 dist_x = x1 > x2 ? x1 - x2 : x2 - x1;
+    UInt64 dist_y = y1 > y2 ? y1 - y2 : y2 - y1;
```
Comment (yariks5s) [inline src/Common/HilbertUtils.h:118]: Maybe it's better to use abs()?

**Comment by Artemmm91** [comment]
@yariks5s hey, i think am ready with PR, have added docs and tests. CI looks nice, the one test that have failed seems not correlated with me

**Comment by yariks5s** [comment]
@Artemmm91 Hi, I was just reviewing this PR, it looks nice and beneficial! I hope we'll merge it soon

**Comment by Artemmm91** [comment]
@alexey-milovidov as a continuation to comment about minmax indexes instead of index analysis(https://github.com/ClickHouse/ClickHouse/pull/55642#issuecomment-1763671198)

i benched on ranged queries (both on ClientIP+RemoteIP) and in two cases - on tables with granules of 8192 rows and granules of 32768 rows

And it seems that index analysis is noticeably faster

Granules of 32768
big_ranges_grand_granules_hilbert_index

Granules of 8192
big_ranges_hilbert_index

**Comment by alexey-milovidov** [comment]
@Artemmm91, This is a really good result! I've seen cases when index analysis was slower on its own, so it was not obvious to me.

**Review by yariks5s — APPROVED** [review]

**Comment by yariks5s** [comment]
@Artemmm91 Could you merge with the master, please?

**Comment by Artemmm91** [comment]
@yariks5s hi! checks have finished, and CI looks normal, seems like one failed check is not correlated

**Comment by yariks5s** [comment]
@Artemmm91 yes, it looks good, thank you!

**Comment by alexey-milovidov** [comment]
Video: https://www.youtube.com/live/BK-x8lpvOQw?si=PKhIRIYMwZ0TRze4&t=2370

**Pull request merged**